### PR TITLE
Remove redundancy in ring morphisms etc

### LINKED ISSUE #1539: Algebra.Morphism.Structures: redundant fields in ring-like morphisms
Consider the `IsRingHomomorphism` structure. It winds up having two different, not-necessarily-equal definitions of `IsRelHomomorphism _≈₁_ _≈₂_ ⟦_⟧`, one via the proof that addition is a group homomorphism and one via the proof that multiplication is a monoid homomorphism.

I'm working on a fix for this that defines, for example, a ring homomorphism as a semiring homomorphism plus a proof that negation is homomorphic.

### PR DESCRIPTION
**Pull request by Taneb**
This redefines NearSemiring, Semiring, and Ring homomorphisms so that
they each build up by adding a single new property to a smaller
structure. This ensures there's only one proof that they're a Rel
homomorphism. For the same reason, Lattice homomorphisms are defined
directly rather than showing it's two magma homomorphisms.

I've tried to make it so that each of these structures provides any
smaller structure at its top level, so if we have a RingMonomorphism we
can easily say that it's a MagmaHomomorphism by multiplication or
whatever, but I might have missed one or two.

Closes #1539

### CODE REVIEW

File: src/Data/Rational/Properties.agda [hunk 1]
```diff
@@ -1068,10 +1085,36 @@ toℚᵘ-isMonoidMonomorphism-* = record
...
+  { isSemiringHomomorphism = toℚᵘ-isSemiringHomomorphism-+-*
+  ; injective              = toℚᵘ-injective
+  }
+
 toℚᵘ-isRingHomomorphism-+-* : IsRingHomomorphism +-*-rawRing ℚᵘ.+-*-rawRing toℚᵘ
 toℚᵘ-isRingHomomorphism-+-* = record
-  { +-isGroupHomomorphism  = toℚᵘ-isGroupHomomorphism-+
-  ; *-isMonoidHomomorphism = toℚᵘ-isMonoidHomomorphism-*
+  -- { +-isGroupHomomorphism  = toℚᵘ-isGroupHomomorphism-+
+  --; *-isMonoidHomomorphism = toℚᵘ-isMonoidHomomorphism-*
```
Comment (MatthewDaggitt) [inline src/Data/Rational/Properties.agda:1115]: Comments can be removed?

**Review by MatthewDaggitt — COMMENTED** [review]
Looks like a much better design, thank you! Only thing missing is a CHANGELOG entry :smile:

**Comment by MatthewDaggitt** [comment]
@Taneb given the number of great PRs recently, I've also added you to the list of Agda developers. You should now have permission to create PRs on this repo directly if you want to :+1:

File: src/Data/Rational/Properties.agda (hunk 1 above)
Comment (Taneb) [reply src/Data/Rational/Properties.agda:1115]: Oops, didn't mean to leave them in

**Comment by Taneb** [comment]
I'm going to make a subsequent PR adding modules like `Algebra.Morphism.NearSemiringMonomorphisms` and proving that Rationals satisfy some more intermediate algebraic structures, because I've added some of the `RawX` bundles, but this PR is already bigger than I set out to do so it can wait a little bit.

**Pull request merged**